Introduce postMessage mechanism for PDF.js viewer and use it for context menu items

nt1m opened the pull request:
#### 5b3fe7dc3092485e4822fb5ae51cb9f80e7ea0de
<pre>
Introduce postMessage mechanism for PDF.js viewer and use it for context menu items
<a href="https://bugs.webkit.org/show_bug.cgi?id=236668">https://bugs.webkit.org/show_bug.cgi?id=236668</a>
&lt;rdar://88985691&gt;

Reviewed by Brent Fulgham.

This introduces a mechanism to postMessage to the PDF.js iframe, and uses it for:
- context menu items
- rendering the initial PDF

Parts of the code by Ryan Reno.

* Source/WebCore/Modules/pdfjs-extras/content-script.js:
(const.PDFJSContentScript.setPageMode):
(const.PDFJSContentScript.open):
(const.PDFJSContentScript.init):
* Source/WebCore/html/PDFDocument.cpp:
(WebCore::PDFDocument::postMessageToIframe):
(WebCore::PDFDocument::sendPDFArrayBuffer):
* Source/WebCore/html/PDFDocument.h:
* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::ContextMenuController::contextMenuItemSelected):
(WebCore::ContextMenuController::performPDFJSAction):
* Source/WebCore/page/ContextMenuController.h:

Canonical link: <a href="https://commits.webkit.org/253223@main">https://commits.webkit.org/253223@main</a>
</pre>

saambarati commented on Source/WebCore/html/PDFDocument.cpp:
```diff
@@ -181,32 +181,48 @@ void PDFDocument::finishedParsing()
+        auto returnValue = contentWindow->postMessage(*contentWindowGlobalObject, *contentWindow, message, "*"_s, Vector { dataRef });
+        if (returnValue.hasException())
+            returnValue.releaseException();
+    }
```
Nit: I think it'd be a little cleaner to not duplicate the exception logic, maybe something like:

```
    WindowPostMessageOptions options;
    if (data)
        options = WindowPostMessageOptions { "*"_s, JSC::Strong<JSC::JSObject> { vm, data }  };
    auto returnValue = contentWindow->postMessage(*contentWindowGlobalObject, *contentWindow, message, WTFMove(options));
    if (returnValue.hasException())
        returnValue.releaseException();
 ```

nt1m replied:
That's much cleaner, thanks!

saambarati commented on Source/WebCore/html/PDFDocument.cpp:
```diff
@@ -181,32 +181,48 @@ void PDFDocument::finishedParsing()
+    JSC::JSLockHolder lock(vm);
+
+    JSC::JSObject* message = constructEmptyObject(globalObject);
+    message->putDirect(vm, JSC::Identifier::fromString(vm, "message"_s), JSC::jsNontrivialString(vm, name));
```
nit: You can use the "message" identifier that's already baked inside CommonIdentifiers on the vm:

```
vm.propertyNames->message
```

saambarati commented on Source/WebCore/html/PDFDocument.cpp:
```diff
@@ -181,32 +181,48 @@ void PDFDocument::finishedParsing()
+    JSC::JSObject* message = constructEmptyObject(globalObject);
+    message->putDirect(vm, JSC::Identifier::fromString(vm, "message"_s), JSC::jsNontrivialString(vm, name));
+    if (data)
+        message->putDirect(vm, JSC::Identifier::fromString(vm, "data"_s), data);
```
nit: Maybe add this identifier to WebCoreBuiltinNames?

nt1m replied:
I want to avoid sprinkling PDF.js-only code across WebCore for now.

saambarati reviewed (commented):
JS part looks good to me, just some nits/suggestions.

pxlcoder commented on Source/WebCore/page/ContextMenuController.cpp:
```diff
@@ -1565,6 +1583,18 @@ void ContextMenuController::showImageControlsMenu(Event& event)
+
+#if ENABLE(PDFJS)
+
+void ContextMenuController::performPDFJSAction(Frame* frame, const String& action)
```
This should take a reference to `Frame` – there is a null check at the top of `ContextMenuController::contextMenuItemSelected`.

nt1m replied:
Done, thanks!

cdumez commented on Source/WebCore/html/PDFDocument.cpp:
```diff
@@ -181,32 +181,40 @@ void PDFDocument::finishedParsing()
+
+    WindowPostMessageOptions options;
+    if (data)
+        options = WindowPostMessageOptions { "*"_s, Vector { JSC::Strong<JSC::JSObject> { vm, data } } };
```
Should probably use starAtom() instead of `"*"_s`.

cdumez replied:
Why are we passing `data` via WindowPostMessageOptions's transferrable? Why isn't it enough to pass them as part of `message.data` ?

Also, why set the target to "*" when there is data but not when there isn't?

nt1m replied:
> Why are we passing data via WindowPostMessageOptions's transferrable? Why isn't it enough to pass them as part of message.data ?

We're passing an ArrayBuffer, so this is better for memory.

> Also, why set the target to "*" when there is data but not when there isn't?

Switched to "/"_s

brentfulgham commented on Source/WebCore/page/ContextMenuController.cpp:
```diff
@@ -548,6 +552,20 @@ void ContextMenuController::contextMenuItemSelected(ContextMenuAction action, co
+        break;
+    case ContextMenuItemPDFTwoPagesContinuous:
+        performPDFJSAction(*frame, "context-menu-two-pages-continuous"_s);
+        break;
```
Nice!

brentfulgham commented on Source/WebCore/page/ContextMenuController.cpp:
```diff
@@ -1565,6 +1583,18 @@ void ContextMenuController::showImageControlsMenu(Event& event)
+void ContextMenuController::performPDFJSAction(Frame& frame, const String& action)
+{
+    auto* pdfDocument = dynamicDowncast<PDFDocument>(frame.ownerElement()->document());
+    ASSERT(pdfDocument);
```
This should either be RELEASE_ASSERT, or check pdfDocument for nullptr before using.

brentfulgham reviewed (approved):
Looks great! Can't wait to see it get into the build.

cdumez reviewed (approved):
LGTM

nt1m committed:
Introduce postMessage mechanism for PDF.js viewer and use it for context menu items
https://bugs.webkit.org/show_bug.cgi?id=236668
<rdar://88985691>

Reviewed by Brent Fulgham.

This introduces a mechanism to postMessage to the PDF.js iframe, and uses it for:
- context menu items
- rendering the initial PDF

Parts of the code by Ryan Reno.

* Source/WebCore/Modules/pdfjs-extras/content-script.js:
(const.PDFJSContentScript.setPageMode):
(const.PDFJSContentScript.open):
(const.PDFJSContentScript.init):
* Source/WebCore/html/PDFDocument.cpp:
(WebCore::PDFDocument::postMessageToIframe):
(WebCore::PDFDocument::sendPDFArrayBuffer):
* Source/WebCore/html/PDFDocument.h:
* Source/WebCore/page/ContextMenuController.cpp:
(WebCore::ContextMenuController::contextMenuItemSelected):
(WebCore::ContextMenuController::performPDFJSAction):
* Source/WebCore/page/ContextMenuController.h:

Canonical link: https://commits.webkit.org/253223@main
